README.md
## Taijilang
Taijilang is a new powerful general programming language. Taijilang have more  extensibility and customizability than all other existing programming languages. 

In general, the features of Taijilang can be summed up: more useful macro than in lisp, more important white space than in python, more general preprocess than in C, more powerful meta compilation than in metalua, more flexible dialect than in rebol, more optimized object code than in coffee-script. 

### Features
* Compiled to javascript language 
* Meta language, and lisp style macro is implemented as a special case of meta language
* Flexible friendly syntax with attaching great importance to indent and white spaces,   
* The most customizable and extensible language, with the help of dynamic parser, macro and meta language features   
* Everything are expressions 
* Module and package
* Optimized object code

### Project Address: 
[github: github.com/taijiweb/taijilang](https://www.github.com/taijiweb/taijilang)

[npm: npmjs.org/package/taiji](https://www.npmjs.org/package/taiji)

[google groups: taijilang](https://groups.google.com/forum/#!forum/taijilang)
 
[G+: taiji lang](https://plus.google.com/u/0/114446069949044102399/posts/p/pub)

QQ群: 太极语言 194928684

### Creator
The creator of taijilang is [曹星明( Caoxingming, assumed name 太极真人(Taiji Zhenren) ) taijiweeb@gmail.com](taijiweeb@gmail.com).

### Thanks
Thanks to the people who make the great things, I learned so much from them:

[John McCarthy](http://www-formal.stanford.edu/jmc/), he invented the great language - [Lisp](http://en.wikipedia.org/wiki/Lisp), Taijilang follows the spirit of lisp in the heart.

[Jeremy Ashkenas](http://ashkenas.com/), he [created coffee-script](https://github.com/jashkenas/coffeescript). I used [coffee-script](http://coffeescript.org/) to write Tajilang and learned a lot from it on how to create an elegant language and compile it to javascript.

[Ryan Dahl](http://www.youtube.com/watch?v=ztspvPYybIY), he created [node.js](http://nodejs.org/), Taijilang runs on node.js.

[Santosh Rajan](http://santoshrajan.com/), he created [lispyscript](github.com/santoshrajan/lispyscript), from which I got the instant spark to creating Taijilang.

[TJ Holowaychuk](http://tjholowaychuk.com/), he shines the two letter "TJ" because of [so many great stuffs he have made](https://github.com/visionmedia), and I am honored to use .tj as the extension name of taijilang code file.

I need thank to many other people and the stuffs made by them, too, although I can not list everyone of them here.
 
At last, I'll thank to my family. 

**Let it go, let's start the game of throne, write a song of ice and fire in the taiji language.**


history.md
2014
2014-11
11-29
  nextToken do not use token.next, use cursor2token instead, more robust

11-13 pass several simple compile tests
11-9 
 gramatical element expression pass all 494 tests (testoperator, testbasic, testparser)
   # add gramatical element expression: [], {}, (), binary!, prefix!, suffix!
   # experiment in trying to add indentBlock!(block!), line!, indentExpression, colonLeadClause!, clauseWithIndentBlock!, but found that does not seem to a good idea.
11-8 
  gramatical element expressoin pass many tests (all of testoperator.coffee)
  pass all tests in testbasic.coffee
11-4
  global tokenType: pass all original 488 tests!  11-1 for statement, var initialization, if-then, while-else, try-catch
11-3 
  import! and export! statement, letloop! statement
  pass the original letloop! statement tests(1 tests)
  fix bug in parsing while! statement
  now pass all 488 tests for parser (except dynamic parsing)
11-2 
  switch-case, let, parser.line, label
  pass 472 tests, only letloop!, import!, export! do not be tested.

2014-10
  10-20
    interpolate string that starts with $: identifier.identifier...identifier
    interpolate string that starts with $: identifier.identifier...identifier
    can not use matchToken in string, so use a simpler definition.
  10-14
    pass more tests of expression
  10-13 
    compact clause expression pass several tests: a.b.c, `.^1

2014-9
  9-28->10-1
  refactor parser for better performace and better design
  space token under new lexical analysis and parser based on token

  it's coded that the framework of new lexical analysis and parser based on token
  space token and pass some tests

  9-1: 
    regexp: /.../  --> /!.../
    parse {} as [hash!]

2014-8
  8-19-8-31 personal private affairs
  8-10 -> 8-18 the definitive guide to taiji language -- taiji zhenjing
  8-5
    ? clause then body
    more preprocess clause
  8-4
    transform string to symbol by escaping it. e.g.: \"x...", \"...x", \"..."

2014-7
  7-31
    ":" at the end of line can replace 'then'
    embedded meta compilation: #call, by evaluating the embedded meta code while running
  7-30
    refactor definition ->, =>, etc,  now they produces [->, [params], oneStatement], instead of statement list
    ellipsis subscript: a[1..3]; b[1...4]; a[...], a[..], x = a[..5]; x = a[...5]
    \ lead symbol to escape it, e.g. \=, \+=, \>>, \/, etc, useful to write macros.
    1...5, 1..5
  7-29
    {x, y, z} = object
    class is implemented
  7-28
    forIn!, forOf!, forIn!!, forOf!!, jsForIn!
  7-27
    fix bug: multi level quasiquote and unquote
    replace row with column(thank to HackWaly)
  7-25 -> 7-26 
    try to implementing class by meta language
  7-20 -> 7-24
    reimplement module after reimplementing meta compilation 
  7-18 
    found a group of new meta operator #-, #&, #&=
  7-17
    (x, @y...) ->, (x, @y...) =>
    refactor meta compilation once again, no metaResultList and assign when metaConvert
  7-16
    refactor meta compilation; no -=>, ==> or macro, macro is just meta compilation
    fix a bug in default parameter, should assign based on whether the paramerter is undefined
  7-15
    do some marketing for taijilang, https://github.com/taijiweb/taijilang got 15 stars.
  ------------------------------------------------------------------------------------------  
  7-14
    taijilang 0.1.0 is released
  7-13
    \-=> and,\-> and \=> becomes |-=> and, |-> and |=>
    npm publish taiji 0.1.0
    recreate git repository, commit and push to github\taijiweb\taijilang
    prepare to release 0.1.0
  7-12
    distinct -> and =>, in => this become self, similar to coffee-script
    default parameter
    call with ellipsis  arguments, e.g. a(x..., y...)
    fix bug in [1 2]
      fix bug in [1 2]:[1 2] is not compiled to [[1, 2]] any more, [1, 2] instead

  7-11
    construct list with ellipsis operator
      x = [x..., y..., z]
    destructive list assign
      [a, b] = x
      [a..., b] = x
      [a, b...] = x
      [a, b..., c] = x
      [x, y] = [1, 2]
    use #= to do meta assignment for macro or meta value
  7-10
    use @@ access outer scope var
    refactor testcommand.coffee
      define parse, compile, run = (outputPath?, filePath) ->
    let [] becomes list, {} becomes expression block, {. .} becomes hash
    let, if, while, for, switch, try, etc do'nt put conjunction in the parsed expression
  7-9 
    prepare to release
  7-8
    ensure new var name does not conflict with existed names
    refactor meta to become a member of env
  7-1
    trying to add module once again

2014-6
  6-21 -> 6-30: refactor transform.coffee: everything is expression
  6-20
    compile files in taitilang
    fix bug in assign fn = (x, y) -> ...
  6-19: 
    fix bug: ellipsis parameter
    formatTaijiJson: pretty print intermediate result in taiji
  6-16: 
    fix a bug: operatorDict['toString'] == Object.toString function
    fix a gotcha: x!=1: 
      add a line in makeIdentifierFn = (charSet, firstCharSet=firstIdentifierCharSet) ->
      ...
      if text[cursor-1]=='!' and text[cursor]=='=' then cursor--
      ...

  6-15 
    finish the robust and modular parser
  6-12
    pass more tests
  6-9 -> 6-11
    under new no token parser, pass some tests about operator, clause, etc.
  6-8 
    remove a lot of package which are not used from node_modules
  6-5 -> 6-7
    pass some basic tests, but the progress is slow, because I'm having a a bad influenza and a terrible headache these days.
  6-4: 
    work for the robustmess and modularity of parser, once again
    remove 150 lines of code in the morning
  6-3 
    x... ellipsis parameter for function: (a, x...), (a, x..., b), (x..., b)
  6-2 
    compile square.tj: demo macro
    compile let.tj
  6-1
    fix bug in transform ['call', caller, args]: need flatten ['begin'...]
    compile nodeserver.tj

2014-5
  compiler, compiling time computation, parsing time computation
  5-30 -> 5-31
    compile chatserver.tj

lib/constant.js
var SYMBOL, symbol;

SYMBOL = 4;

exports.symbol = symbol = function(value) {
  return {
    value: value,
    kind: SYMBOL
  };
};

module.exports = {
  OPERATOR_EXPRESSION: 0,
  COMPACT_CLAUSE_EXPRESSION: 1,
  NULL: 0,
  NUMBER: 1,
  STRING: 2,
  IDENTIFIER: 3,
  SYMBOL: SYMBOL,
  REGEXP: 5,
  HEAD_SPACES: 6,
  CONCAT_LINE: 7,
  PUNCTUATION: 8,
  FUNCTION: 9,
  BRACKET: 10,
  PAREN: 11,
  CURVE: 13,
  INDENT_EXPRESSION: 14,
  NEWLINE: 15,
  SPACES: 16,
  INLINE_COMMENT: 17,
  SPACES_INLINE_COMMENT: 18,
  LINE_COMMENT: 19,
  BLOCK_COMMENT: 20,
  CODE_BLOCK_COMMENT: 21,
  CONCAT_LINE: 22,
  MODULE_HEADER: 23,
  MODULE: 24,
  NON_INTERPOLATE_STRING: 25,
  INTERPOLATE_STRING: 26,
  INDENT: 27,
  UNDENT: 28,
  HALF_DENT: 29,
  EOI: 30,
  C_BLOCK_COMMENT: 31,
  SPACE_COMMENT: 32,
  TAIL_COMMENT: 33,
  SPACE: 34,
  HASH: 35,
  RIGHT_DELIMITER: 36,
  KEYWORD: 37,
  CONJUNCTION: 38,
  CODE_BLOCK_COMMENT_LEAD_SYMBOL: 39,
  PREFIX: 40,
  SUFFIX: 41,
  BINARY: 42,
  VALUE: 43,
  UNDEFINED: symbol('undefined'),
  BEGIN: symbol('begin!'),
  IF: symbol('if'),
  PREFIX: symbol('prefix!'),
  SUFFIX: symbol('suffix!'),
  BINARY: symbol('binary!'),
  WHILE: symbol('while'),
  BREAK: symbol('break'),
  CONTINUE: symbol('continue'),
  THROW: symbol('throw'),
  RETURN: symbol('return'),
  NEW: symbol('new'),
  FORIN: symbol('forIn!'),
  FOROF: symbol('forOf!'),
  JSFORIN: symbol('jsForIn!'),
  TRY: symbol('try'),
  SHARP: symbol('#'),
  PAREN_PAIR: symbol('()'),
  BRACKET_PAIR: symbol('[]'),
  CURVE_PAIR: symbol('{}'),
  LABEL: symbol('label!'),
  VAR: symbol('var'),
  CALL: symbol('call!'),
  ATTR: symbol('attribute!'),
  INDEX: symbol('index!'),
  FUNCTION: symbol('function'),
  LIST$: symbol('list!')
};
